src/main/java/org/junitpioneer/jupiter/StdIoExtension.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static java.lang.String.format;
+
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.lang.reflect.Method;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+class StdIoExtension implements ParameterResolver, BeforeTestExecutionCallback, AfterEachCallback {
+
+	static final String SEPARATOR = System.getProperty("line.separator");
+
+	private static final Namespace NAMESPACE = Namespace.create(StdIoExtension.class);
+
+	private static final String SYSTEM_IN_KEY = "StdIo_System_In";
+	private static final String STD_IN_KEY = "StdIo_Std_In";
+	private static final String OUT_KEY = "StdIo_Std_Out";
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		Class<?> type = parameterContext.getParameter().getType();
+		return (type == StdIn.class || type == StdOut.class);
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		Class<?> parameterType = parameterContext.getParameter().getType();
+		if (parameterType == StdOut.class)
+			return prepareStdOut(extensionContext);
+		if (parameterType == StdIn.class) {
+			String[] source = extensionContext.getRequiredTestMethod().getAnnotation(StdIo.class).value();
+			if (source.length == 0)
+				throw new ExtensionConfigurationException(
+					"@StdIo defined no input, so System.in is still in place and no StdIn parameter can be provided. "
+							+ "If you want to define empty input, use `@StdIo(\"\")`.");
+			else
+				//@formatter:off
+				return extensionContext
+						.getStore(NAMESPACE)
+						.getOrComputeIfAbsent(
+								STD_IN_KEY,
+								__ -> createSwapStoreStdIn(extensionContext, source),
+								StdIn.class);
+				//@formatter:on
+		}
+		throw new ParameterResolutionException(format("Could not resolve parameter of type %s.", parameterType));
```

Comment:
This exception is never thrown. Could be deleted.
From the [ParameterResolver::resolveParameter documentation](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/extension/ParameterResolver.html#resolveParameter-org.junit.jupiter.api.extension.ParameterContext-org.junit.jupiter.api.extension.ExtensionContext-):
> This method is only called by the framework if supportsParameter(org.junit.jupiter.api.extension.ParameterContext, org.junit.jupiter.api.extension.ExtensionContext) previously returned true for the same ParameterContext and ExtensionContext.

Reply:
Without this line, the compiler complains because the method does not return/throw on all paths. I prefer an exception over `return null`.